#  Part 1.  Calculer avec Python 3.x

Quelle est la différence entre Python et une calculette ? On commence cette leçon en montrant comment Python peut servir de **calculette**, puis nous verrons quelques éléments de base en programmation : types de données, variables, listes, et boucles.
What is the difference between *Python* and a calculator?  We begin this first lesson by showing how Python can be used **as** a calculator, and we move into some of the basic programming language constructs:  data types, variables, lists, and loops.

## Table des matières

- [Python comme calculette](#calculator)
- [Calculer avec des booléens](#booleans)
- [Déclarer des variables](#variables)
- [Ranges](#ranges)
- [Iteration sur un range](#iterating)

<a id='calculator'></a>

## Python comme calculette

Différentes sortes de données sont stockées avec des types différents en Python. Par exemple, si vous voulez travailler avec des entiers, vos données seront stockées avec type *int*. Un nombre réel doit être stocké en *float*. Il y a aussi les types booléens (True/False), les chaines de caractères (comme "Hello World!"), et beaucoup d'autres types.

On trouvera une référence complète pour les types numériques et les opérations arithmétiques dans la [documentation  Python officielle](https://docs.python.org/2/library/stdtypes.html#numeric-types-int-float-long-complex).

Les opérations arithmétiques en Python sont  dénotés `+`, `-`, `*`, and `/`. Evaluez chacune des cellules suivantes pour voir comment Python traite ces opérations sur les *entiers*. A mesure que vous avancez dans cette leçon et les suivantes, essayez de *prévoir* ce que produira un calcul avant de l'effectuer.  

In [6]:
2 + 3

5

In [7]:
2 * 3

6

In [8]:
5 - 11

-6

In [9]:
5.0 - 11

-6.0

In [10]:
5 / 11

0.45454545454545453

In [11]:
6 / 3

2.0

In [12]:
5 // 11

0

In [13]:
6 // 3

2

Les résultats ne doivent pas être trop surprenants, mais les deux derniers demandent une petite explication. Python *interprète* le nombre 5 comme un *int* (entier) et 5.0 comme un *float* (décimal).

Python permet deux sortes de division.  En Python 3.x une division avec un seul signe slash donne un résultat *float* ; avec deux signes slash on obtient le quotient *int* de la division euclidienne.

Ceci est utile mais on doit faire attention. **Ce tutoriel suppose qu'on utilise Python 3.x.** Si vous êtes en Python 2.x, attention : les deux symboles de la division se comportent autrement.

In [5]:
-12 // 5

-3

Python vous permet de grouper des expressions entres parenthèses, et suit les priorités opératoires habituelles.

In [15]:
(3 + 4) * 5

35

In [16]:
3 + (4 * 5)

23

In [17]:
3 + 4 * 5   #  Prévoir le résultat avant d'exécuter

23

Ci-dessous, une cellule vide pour vous exercer. Rappel : pour créer une nouvelle cellule, faire **Echap a** ou **Echap b**; pour supprimer une cellule faire **Echap x**.

In [18]:
#  Cellule vide ; expérimenter.


En arithmétique, la division euclidienne est très importante. La division entière (deux signes slash) fournit le quotient, le symbole `%` fournit le reste.

In [6]:
23 // 5  # quotient

4

In [7]:
23 % 5  # reste

3

Remarquez ci-dessus le hashtag `#` pour commenter dans une ligne de code.

Python possède aussi une commande unique pour la division euclidienne.  Le résultat est un *tuple*.

In [21]:
divmod(23,5)

(4, 3)

In [22]:
type(divmod(23,5))

tuple

La fonction *type* permet de connaitre le type des données.

In [23]:
type(3)

int

In [24]:
type(3.0)

float

In [25]:
type('Hello')

str

In [26]:
type([1,2,3])

list

Il est très important de connaitre letype des données que l'on manipule, et comment Python va opérer sur ces différents types.

In [27]:
3 + 3

6

In [28]:
3.0 + 3.0

6.0

In [29]:
'Hello' + 'World!'

'HelloWorld!'

In [30]:
[1,2,3] + [4,5,6]

[1, 2, 3, 4, 5, 6]

In [31]:
3 + 3.0

6.0

In [32]:
3 + 'Hello!'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [33]:
#  Cellule vide, pour s'exercer.


Comme vous pouvez le constater, l'addition a un sens entre valeurs numériques ou entre chaines de caractères, mais pas entre un nombre et une chaine.

Par ailleurs, la multiplication a un sens entre un entier et une chaine ou une liste.

In [34]:
3 * 'Hello!'

'Hello!Hello!Hello!'

In [35]:
0 * 'Hello!'

''

In [36]:
2 * [1,2,3]

[1, 2, 3, 1, 2, 3]

Pouvez vous créer une chaine de 100 A's (comme `AAA...`)?  Essayez dans la cellule ci-dessous.

In [ ]:
# créer ici une chaine de 100 A's


Les exposants en Python sont donnés par l'opérateur `**`.

In [9]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [39]:
2.0**1000

1.0715086071862673e+301

Comme tout-à-l'heure, Python interprète l'operation (`**`) differéremment suivant les types. Avec des types entiers, le calcul est **exact**. Une bonne nouvelle, en Python, c'est que les entiers peuvent être de longueur arbitraire, comme ci-dessus, contrairement à des langages comme C/C++.

La notation e+301 signifie "multiplié par `10**301`"; cela s'appelle *notation scientifique*.

In [40]:
type(2**1000)

int

In [41]:
type(2.0**1000)

float

### Exercises

1.  Quels types avez vous vus, et sur quelles données sont ils utilisés ? Donner des exemples.

2.  Comment se comporte la division `/` sur differents types ?

3.  Même question avec la multiplication `*`.

4.  Quelle est la différence entre 100 et 100.0, pour Python?

Double-cliquez cette cellule markdown pour l'éditer, et répondre aux questions de l'exercice.

<a id='booleans'></a>

## Calculer avec des booléens

Un booléen (type *bool*) est le plus petit morceau de donnée possible ; il ne prend que les valeurs  *True* or *False*. C'est en particulier utile pour tester des égalités ou inégalités entre nombres. Ci-dessous quelques exemples. Voir aussi [la documentation Python officielle](https://docs.python.org/2/library/stdtypes.html#boolean-operations-and-or-not).

In [43]:
3 > 2

True

In [44]:
type(3 > 2)

bool

In [45]:
10 < 3

False

In [46]:
2.4 < 2.4000001

True

In [47]:
32 >= 32

True

In [48]:
32 >= 31

True

In [49]:
2 + 2 == 4

True

Qui est le plus grand:  $23^{32}$ or $32^{23}$?  Utilisez la cellule ci-dessous pour répondre !

In [50]:
#  Write your code here.


Les expressions `<`, `>`, `<=`, `>=` arsonte interprétées comme des **operations** avec un input numerique et un output booléen.  Le symbole `==` teste si deux nombres sont égaux.  Attention à ne pas confondre avec `=`qui a une signification toute différente (déclaration de variable, voir plus bas).

Using the remainder operator `%` and equality, we obtain a divisibility test.

In [51]:
63 % 7 == 0  # 63 est-il divisible par 7 ?

True

In [52]:
101 % 2 == 0  # 101 est-il pair ?

False

Utilisez la cellule dessous pour déterminer si 1234567890 est divisible par 3.

In [53]:
# Votre code ici.


Entre booléens on peut utiliser les opérateurs and, or, not.

Ci-dessous les **tables de vérité** des opérateurs **and** et **or**.

| and || True | False |
|-----||------|-------|
|||||
| **True** || True | False |
| **False** || False | False|

| or || True | False |
|-----||------|-------|
|||||
| **True** || True | True |
| **False** || True | False|


In [54]:
True and False

False

In [55]:
True or False

True

In [56]:
True or True

True

In [57]:
not True

False

Utilisez les tables de vérité pour prévoir le résultat des expressions suivantes, avant d'évaluer le code.

In [58]:
(2 > 3) and (3 > 2)

False

In [59]:
(1 + 1 == 2) or (1 + 1 == 3)

True

In [60]:
not (-1 + 1 >= 0)

False

In [61]:
2 + 2 == 4

True

In [10]:
2 + 2 != 4  # Pour "not equal", Python utilise l'opérateur `!=`.

False

In [63]:
2 + 2 != 5  

True

In [64]:
not (2 + 2 == 5)

True

Python accepte quelques opérations entre booléens et nombres. Expérimentez ci-dessous et changez les valeurs.

In [13]:
False * 100

0

In [67]:
True + 13

14

L'aptitude de Python à interpréter les opérations suivant le contexte a ses avantages et inconvénients. D'un côté ça permet des codes plus courts, de l'autre ça peut diminuer la lisibilité du code. Les bons programmeurs visent à produire du code pas seulement court, mais aussi lisible.

Le [Zen of Python](https://www.python.org/dev/peps/pep-0020/) est une série de 20 aphorismes pour programmeurs Python.  Voici les sept premiers.

> Beautiful is better than ugly.

> Explicit is better than implicit.

> Simple is better than complex.

> Complex is better than complicated.

> Flat is better than nested.

> Sparse is better than dense.

> Readability counts.

### Exercises

3.  Comment savoir, en une ligne de code, si 2349872348723, est un multiple de 7 mais pas un multiple de 11 ?

6.  L'opérateur `xor` veut dire "exclusive or".  Sa table de vérité est : `True xor True = False` et `False xor False = False` et `True xor False = True` et `False xor True = True`.  Implémenter `xor` à l'aide des opérateurs `and`, `or`, et `not`.

In [14]:
#  Espace pour répondre aux questions de l'exercice

<a id='variables'></a>

## Déclaration de variables

Une caractéristique essentielle en programmation est la déclaration de variables. Quand vous déclarez une variable, vous *stockez* une donnée en mémoire et vous donnez un *nom* à cette donnée. Tout ceci se fait à l'aide du symbole `=`.

In [69]:
e = 2.71828

In [70]:
e * e

7.3890461584

In [71]:
type(e)

float

Un nom de variable doit commencer avec une lettre, ne doit pas contenir d'espaces, et ne doit pas être un mot Python existant (comme `True` par exemple). On peut utiliser des lettres (minuscules et majuscules), des chiffres et des underscores `_`.  

Donc, comme nom de variable, `e` est valide mais `type` ne l'est pas. Les débutants sont tentés d'utiliser des abréviations courtes comme `dx` or `vbn`.  Ce n'est pas recommandé ; utilisez plutôt des noms explicites comme `difference_x` ou `very_big_number`.  Cela fera un code plus lisible pour vous et les autres !

Il y a différentes conventions de style pour les noms de variables. Nous utiliserons des noms en minuscule, et des underscores pour séparer les mots, suivant à peu près les [Google's style conventions](https://google.github.io/styleguide/pyguide.html#Python_Style_Rules) for Python code.

In [72]:
my_number = 17

In [73]:
my_number < 23

True

Après avoir déclaré une variable, sa valeur reste la même jusqu'à ce qu'elle soit modifiée. Vous pouvez modifier la valeur d'une variable avec une assignation simple. Après les lignes ci-dessus, la valeur de my_number est 17.

In [74]:
my_number = 3.14

Cette commande réaffecte la valeur de my_number à 3.14. Notez que cela change aussi le type! Elle remplace effectivement la valeur précédente et la remplace par la nouvelle valeur.

Il est souvent utile de changer la valeur d'une variable *de manière incrémentielle* ou *récursive*. Python, comme de nombreux langages de programmation, permet d’assigner des variables de manière auto-référentielle. Que pensez-vous de la valeur de S après les quatre lignes suivantes?

In [15]:
S = 0
S = S + 1
S = S + 2
S = S + 3
print(S)

6


Considérez maintenant les commandes suivantes

In [76]:
my_number = 17
new_number = my_number + 1
my_number = 3.14

Quelles sont les valeurs de my_number et new_number,après l'exécution de ces trois lignes ?

Pour accéder à ces valeurs vous pouvez utiliser la fonction *print*.

In [77]:
print(my_number)
print(new_number)

3.14
18


### Exercises

1. What is the difference between `=` and `==` in the Python language ?

3. Imaginez deux variables `a` et `b`, et vous voulez permuter leurs valeurs.  Comment faites-vous ?

In [16]:
#  Répondre ici, avec un exemple.

<a id='ranges'></a>

## Lists and ranges

Python se distingue par le rôle central joué par les listes (type `list`). Les données d'une liste peuvent être de tout type. Plusieurs types sont possibles dans la même liste! La syntaxe de base d'une liste consiste à utiliser des crochets pour inclure les éléments de liste et les virgules pour séparer les éléments de la liste.

In [79]:
type([1,2,3])

list

In [80]:
type(['Hello',17])

list

Il y a un autre type appelé un tuple que nous utiliserons rarement (type `tuple`). Les tuples utilisent des parenthèses pour la clôture au lieu des parenthèses.

In [81]:
type((1,2,3))

tuple

Il existe un autre type de liste dans Python 3, appelé range. Les ranges sont un peu comme des listes, mais au lieu de mettre chaque élément dans un emplacement de mémoire, les ranges doivent simplement retenir trois nombres entiers: leur début, leur arrêt et leur pas.  

La commande `range` crée un range avec un début, un arrêt et un pas donnés. Si vous ne saisissez qu'un seul entier en input, le range **commence à zéro** et utilise un **pas de un ** et arrête **juste avant le numéro d'arrêt**.

On peut créer une liste à partir d'un range (en plaçant chaque terme du range dans un emplacement de mémoire), en utilisant la commande `list`. Voici quelques exemples.

In [82]:
type(range(10)) # Ranges are their own type, in Python 3.x.  Not in Python 2.x!

range

In [83]:
list(range(10)) # Let's see what's in the range.  Note it starts at zero!  Where does it stop?

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

A more complicated two-input form of the range command produces a range of integers **starting at** a given number, and **terminating before** another given number.

In [84]:
list(range(3,10))

[3, 4, 5, 6, 7, 8, 9]

In [3]:
list(range(-4,5))

[-4, -3, -2, -1, 0, 1, 2, 3, 4]

C'est une source commune de difficulté pour les débutants en Python. Alors que le premier paramètre (-4) est le point de départ de la liste, la liste se termine juste avant le deuxième paramètre (5). Cela prend un peu de temps pour s'y habituer, mais les programmeurs Python expérimentés ont tendance à aimer cette convention.

La longueur d'une liste est accessible par la commande `len`.

In [86]:
len([2,4,6])

3

In [87]:
len(range(10))  # La commande len peut traiter des listes et des plages. Pas besoin de convertir

10

In [88]:
len(range(10,100)) # Prévoir le résultat avant d'évaluer

90

La variante finale de la commande range (pour l'instant) est la commande à trois paramètres `range(a, b, s)`.
Cela produit une liste comme le range (a, b), mais avec "un pas de s".
En d'autres termes, il produit une liste d'entiers, commençant à a, augmentant de s d'une entrée à l'autre, allant jusqu'à (mais n'incluant pas) b. Il vaut mieux expérimenter un peu !

In [89]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [90]:
list(range(11,30,2))

[11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

In [91]:
list(range(-4,5,3))

[-4, -1, 2]

In [92]:
list(range(10,100,17))

[10, 27, 44, 61, 78, 95]

Cela peut aussi être utilisé pour les ranges décroissants ; observer que le nombre final b dans la plage (a, b, s) n'est pas inclus.

In [93]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

Combien de multiples de 7 sont compris entre 10 et 100? Nous pouvons répondre assez rapidement avec les commandes `range` et `len`.

In [94]:
list(range(10,100,7))  # What list will this create?  It won't answer the question...

[10, 17, 24, 31, 38, 45, 52, 59, 66, 73, 80, 87, 94]

In [95]:
list(range(14,100,7))  # Starting at 14 gives the multiples of 7.

[14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]

In [96]:
len(range(14,100,7))  # Gives the length of the list, and answers the question!

13

### Exercises

1.  Si `a` et `b` sont des entiers, quelle est la longueur de `range(a,b)`?

2.  Combien de multiples de 7 sont compris entre 10 et 100? Vous pouvez répondre assez rapidement avec les commandes `range` et `len`.

3.  Créez la liste `[1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5]` avec la commande range et une autre opération.


In [97]:
#  Espace pour répondre aux questions.


<a id='iterating'></a>

## Iterating over a range

Computers are excellent at repetitive reliable tasks.  If we wish to perform a similar computation, many times over, a computer a great tool.  Here we look at a common and simple way to carry out a repetetive computation:  the "for loop".  The "for loop" *iterates* through items in a list or range, carrying out some action for each item.  Two examples will illustrate.

In [98]:
for n in [1,2,3,4,5]:
    print(n*n)

1
4
9
16
25


In [99]:
for s in ['I','Am','Python']:
    print(s + "!")

I!
Am!
Python!


The first loop, **unraveled**, carries out the following sequence of commands.

In [100]:
n = 1
print(n*n)
n = 2
print(n*n)
n = 3
print(n*n)
n = 4
print(n*n)
n = 5
print(n*n)

1
4
9
16
25


But the "for loop" is more efficient *and* more readable to programmers.  Indeed, it saves the repetition of writing the same command `print n*n` over and over again.  It also makes transparent, from the beginning, the range of values that `n` is assigned to.  

When you read and write "for loops", you should consider how they look unravelled -- that is how Python will carry out the loop.  And when you find yourself faced with a repetetive task, you might consider whether it may be wrapped up in a for loop.

Try to unravel the loop below, and predict the result, before evaluating the code.

In [101]:
P = 1
for n in range(1,6):
    P = P * n
print(P)

120


This might have been difficult!  So what if you want to trace through the loop, as it goes?  Sometimes, especially when debugging, it's useful to inspect every step of the loop to see what Python is doing.  We can inspect the loop above, by inserting a print command within the *scope* of the loop.

In [102]:
P = 1
for n in range(1,6):
    P = P * n
    print("n is",n,"and P is",P)
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Here we have used the *print* command with strings and numbers together.  In Python 3.x, you can print multiple things on the same line by separating them by commas.  The "things" can be strings (enclosed by single or double-quotes) and numbers (int, float, etc.).

In [103]:
print("My favorite number is",17)

My favorite number is 17


If we unravel the loop above, the linear sequence of commands interpreted by Python is the following.

In [104]:
P = 1
n = 1
P = P * n
print("n is",n,"and P is",P)
n = 2
P = P * n
print("n is",n,"and P is",P)
n = 3
P = P * n
print("n is",n,"and P is",P)
n = 4
P = P * n
print("n is",n,"and P is",P)
n = 5
P = P * n
print("n is",n,"and P is",P)
print (P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Let's analyze the loop syntax in more detail.  
```python
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)
```
The "for" command ends with a colon `:`, and the **next two** lines are indented.  The colon and indentation are indicators of **scope**.  The *scope* of the for loop begins after the colon, and includes all indented lines.  The *scope* of the for loop is what is repeated in every step of the loop (in addition to the reassignment of `n`).  

In [105]:
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


If we change the indentation, it changes the scope of the for loop.  Predict what the following loop will do, by unraveling, before evaluating it.

In [106]:
P = 1
for n in range(1,6):
    P = P * n
print("n is",n,"and P is",P)
print(P)

n is 5 and P is 120
120


Scopes can be nested by nesting indentation.  What do you think the following loop will do?  Can you unravel it?

In [107]:
for x in [1,2,3]:
    for y in ['a', 'b']:
        print(x,y)

1 a
1 b
2 a
2 b
3 a
3 b


How might you create a nested loop which prints `1 a` then `2 a` then `3 a` then `1 b` then `2 b` then `3 b`?  Try it below.

In [108]:
# Insert your loop here.

Among popular programming languages, Python is particular about indentation.  Other languages indicate scope with open/close braces, for example, and indentation is just a matter of style.  By requiring indentation to indicate scope, Python effectively removes the need for open/close braces, and enforces a readable style.

We have now encountered data types, operations, variables, and loops.  Taken together, these are powerful tools for computation!  Try the following exercises for more practice.  You can also try the exercises at the end of Chapter 0 of [An Illustrated Theory of Numbers](http://bookstore.ams.org/mbk-105/) -- some can be done easily by writing a few lines of Python code.

## Exercises

1.  Describe how Python interprets division with remainder when the divisor and/or dividend is negative.
2.  What is the remainder when $2^{90}$ is divided by $91$?
3.  How many multiples of 13 are there between 1 and 1000?
4.  How many *odd* multiples of 13 are there between 1 and 1000?
5.  What is the sum of the numbers from 1 to 1000?
6.  What is the sum of the squares, from $1 \cdot 1$ to $1000 \cdot 1000$?